presto-cache/src/main/java/com/facebook/presto/cache/RangeMapCacheManager.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.cache;
+
+import com.google.common.collect.Range;
+import com.google.common.collect.RangeMap;
+import com.google.common.collect.TreeRangeMap;
+import io.airlift.slice.Slice;
+import org.apache.hadoop.fs.Path;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.nio.file.Files;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.google.common.collect.Iterators.getOnlyElement;
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
+import static java.nio.file.StandardOpenOption.APPEND;
+
+/**
+ * A naive and self-contained caching system that does not consider ranges of a file:
+ * - It does not consider memory utilization; flushes every requested data to disk.
+ * - It does not expire cache on disk.
+ */
+public class RangeMapCacheManager
+        extends AbstractAsyncCacheManager
+{
+    @SuppressWarnings("UnstableApiUsage")
+    private final Map<Path, RangeMap<Long, CacheFile>> persisted = new ConcurrentHashMap<>();
```

Comment:
Just want to check if my following understanding is correct:

Each `Path` (this probably would only work for hive or HDFS-like connector?) here means a hive file, and the `RangeMap` is the cache to the hive file. Each `Long` in the `RangeMap` is the offset from the original file, and a continuous range would point to the same `CacheFile`. We will do file merging to reduce the number of `CacheFile`, and if a whole file for this `Path` is cached, we will have all the `Long` within the file length range pointing to the same `CacheFile`.